src/ValidationExemption.php

```diff
@@ -0,0 +1,247 @@
+<?php
+/**
+ * Class ValidationExemption.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP;
+
+use AmpProject\Dom\Document;
+use AmpProject\Dom\Element;
+use DOMAttr;
+use DOMNode;
+
+/**
+ * Helper functionality to deal with validation exemptions.
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.2
+ * @internal
+ * @see \AmpProject\DevMode
+ */
+final class ValidationExemption {
+
+	/**
+	 * HTML attribute to indicate one or more attributes have been verified for PX from AMP validation.
+	 *
+	 * @var string
+	 */
+	const PX_VERIFIED_ATTRS_ATTRIBUTE = 'data-px-verified-attrs';
+
+	/**
+	 * HTML attribute to indicate an tag/element has been verified for PX.
+	 *
+	 * The difference here with `data-amp-unvalidated-tag` is that the PX-verified means that the tag will work
+	 * properly Bento components and CSS tree shaking.
+	 *
+	 * @var string
+	 */
+	const PX_VERIFIED_TAG_ATTRIBUTE = 'data-px-verified-tag';
+
+	/**
+	 * HTML attribute to indicate one or more attributes are exempted from AMP validation.
+	 *
+	 * @var string
+	 */
+	const AMP_UNVALIDATED_ATTRS_ATTRIBUTE = 'data-amp-unvalidated-attrs';
+
+	/**
+	 * HTML attribute to indicate an tag/element is exempted from AMP validation.
+	 *
+	 * @var string
+	 */
+	const AMP_UNVALIDATED_TAG_ATTRIBUTE = 'data-amp-unvalidated-tag';
+
+	/**
+	 * Check whether PX is verified for node.
+	 *
+	 * This means that it is exempted from AMP validation, but it doesn't mean the PX is negatively impacted.
+	 *
+	 * @param DOMNode $node Node.
+	 * @return bool Whether node marked as being PX-verified.
+	 */
+	public static function is_px_verified_for_node( DOMNode $node ) {
+
+		// Check element.
+		if (
+			$node instanceof Element
+			&&
+			$node->hasAttribute( self::PX_VERIFIED_TAG_ATTRIBUTE )
+		) {
+			return true;
+		}
+
+		// Check attribute.
+		if (
+			$node instanceof DOMAttr
+			&&
+			self::check_for_attribute_token_list_membership( $node, self::PX_VERIFIED_ATTRS_ATTRIBUTE )
+		) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Mark node as being PX-verified.
+	 *
+	 * @param DOMNode $node Node.
+	 * @return bool Whether successful.
+	 */
+	public static function mark_node_as_px_verified( DOMNode $node ) {
+
+		if ( $node instanceof Element ) {
+			if ( ! $node->hasAttribute( self::PX_VERIFIED_TAG_ATTRIBUTE ) ) {
+				if ( ! $node->ownerDocument ) {
+					return false; // @codeCoverageIgnore
+				}
+				$node->setAttributeNode( $node->ownerDocument->createAttribute( self::PX_VERIFIED_TAG_ATTRIBUTE ) );
+			}
+			return true;
+		}
+
+		if ( $node instanceof DOMAttr ) {
```

Comment:
If the `$node` was already detected to be an instance of `Element`, there's no point in checking it for `DOMAttr`.

```suggestion
		} elseif ( $node instanceof DOMAttr ) {
```